### DISCUSSION TITLE
Types of documentation where it belongs

### DISCUSSION THREAD

**Discussion by julie777** [pyscaffold/pyscaffold#604] [Dev Discussions]
There are many types of documentation or documents relating to a project. Some of them have obvious places. Others just end up somewhere. Sadly some disappear or are impossible to find. I would like to discuss how we manage all of our documents. The table below should give you a starting point to think about this. This also provides motivation for #602.

| Kind      | Section |  Audience  | Source At  |  Published At  |  Purpose |  
|  :----------    | :-- |  :----------  |  :------------  |  :--------------  |  :--------- |
|  Introduction  |  README.txt  | potential users |  project root  |  GitHub and PyPi Project Pages  |  Provide an overview of the project so that potential users can decide if it meets their needs. Provide references to other documentation |
|  User Manual   | Overview | users  |  project/docs   |  readthedocs.io or product website  |  Provide the user with all the information they need to use the product|
|  | Installation |  |  |  | |
|  | Configuration |  |  |  | |
|  | Usage |  |  |  | |
|  |  Tips          |  |  | wiki prefered because users can add to it | |
|  | Command Reference |  |  |  | |
|  | API Reference |  |  |  | |
| product issues | bugs | "users, developers" |   |  github issues  |  provide a way for users to submit bug reports or enhancement requests. Allow developers to manage changes |
|  | enhancements |  |  | issues  | |
| discussions  | general | users |   | github discussions | any product discussion that doesn't fit in another category|
|                | questions | users |   |  discussions |  allow users to ask how to questions and get help|
|                | features | "users, contributors" |   |  discussions |  Allow users and developers to discuss product features |
|                |  |   |   |  discussions | |
|                | development | contributor |   |  discussions |  Allow developers to discuss development topics |
|   | planning | contributor |   |  discussions |  Document the plan for new features or versions of the product |
| Product Documentation | milestones | contributor |   | github wiki |   Document project milestones |
|  | architecture | contributor |   | wiki |  Document the architecture of the product |
|  | design decisions | contributor |   | wiki |  Document project design decisions |
|  | contributing | contributor |  | wiki | how can someone contribute|
|  | revising documentation | contributor |  | wiki | how can a contributor provide changes to the documenation|
|  | making code changes | developer |  | wiki | how can a contributor make code changes|
|  metadata      | pyproject.toml | "developers, tools" |  project/  |   |   provide information necessary to build and deploy project |
|  | setup.cfg |  |  |  | |
|  | tool config |  |  |  | |
|  | version file |  |  |  | |
| code |  | developer | project/src |  | |
| tests |  | "developer, tester" | project/tests |  | |

(I realize that I am enthusiastic and passionate. If I need to slow down or back off just tell me how I can fit it better.)

**Comment by julie777** [comment]
Sadly, my table didn't end up like it did when I previewed it. The preview didn't size columns to wrap in the middle of words.
I need to get better with github markdown.

**Comment by abravalheri** [comment]
Hi @julie777, currently all the topics you listed for the wiki, except `Document project milestones`, are part of a separated section of the website.

Can you help me to understand what would be the main advantages of moving this documentation to the wiki?

**Reply by julie777** [reply]
I'm not sure I was suggesting moving a bunch of stuff, and certainly not immediately. 

I have been using wikis at work for many years. The biggest benefit is the ease of contribution. A user manual like we have can be very nice, but for many things it is much easier to see a spelling error and just fix it. Or try to do something following the instructions and after struggling a bit trying to "get it", you just add a comment for future users.

**Comment by abravalheri** [comment]
(To me a clear advantage of keeping them in the website is that we have all the power of `sphinx` at out hands, while GFM is very limited)

It would be interesting to understand what are the pros/cons of each approach to take an informed decision.

**Comment by julie777** [comment]
BTW, I wasn't trying to specifically suggest a change, but rather start a discussion. Documentation is an extremely important part of a project, and one that is often overlooked. _I am in awe of the current documentation for PyScaffold!_

I do suggest the wiki for developer documentation, capturing the results of discussions, etc. The reasons I have for this comes from many years of experience with developers who don't like to write documentation. They don't capture design decisions. Two developers talk about something, never write it down and months later other developers have to rediscover it.

@abravalheri  The biggest thing I learned when I was the one to deploy the first wiki at a company I worked for 20 years ago, and I think it is absolutely the most important part (when not talking about end users).
> Content is more important than presentation.

To me markdown is just a variation on wiki markup which predated it by many years.

A question to ask yourself, "If I only have 30 minutes to write something down and make sure it is captured, is it better to spend 80% of my time on the content and the rest on the formatting, or do the reverse?"

Or, "do I want it available right now, or next time we build and deploy the doc?"

The wiki makes it easy to add simple formatting, and add links to other sections. (I actually use OneNote for all my note taking and it is much the same. I have at least 20 pages of notes about pyscaffold.) 

Another benefit is notifications of changes, just like discussions and issues. (I hope that works on github. It has on every other wiki I've used)

I don't know if you have worked with a wiki for developer content before, so I am not sure if I am presenting it well. What I would suggest is that we enable it and start using it for some new things, such as any of the changes going in to 5.0. We can see how everyone likes it. 

Trying something new is easy if you don't have to change anything that exists to do it. 😄

**Comment by abravalheri** [comment]
> I do suggest the wiki for developer documentation, capturing the results of discussions, etc. The reasons I have for this comes from many years of experience with developers who don't like to write documentation. They don't capture design decisions. Two developers talk about something, never write it down and months later other developers have to rediscover it.

I like this idea a lot.

Somethings to have in mind:

* There is a very thin line between some users and PyScaffold developers. There are 2 kinds of users:

  - End users that just use the `putup` CLI
  - Users that create extensions

  Right now most of the developer documentation ([how PyScaffold works internally](https://pyscaffold.org/en/latest/dev-guide.html#architecture) for example) is extremely useful for the second kind of users. Equally important: most of the development of PyScaffold itself is done internally via "builtin extensions", which means that the [docs targeting users](https://pyscaffold.org/en/latest/extensions.html) are also very useful to developers. Sphinx allow us to mix and match these parts so they can be presented nicely regardless of the target audience.

* The website is an important resource for PyScaffold users, and it is important to make it easy for them to discover how to contribute. It is also important to make it easy for curious minds to discover how PyScaffold works internally, so they can consider contributing in the future. (Therefore having the 2 types of documentation alongside is a good thing - of course they need to be separated in different sections and have clear titles to avoid confusion).

* Once a development decision is made, the outcome in terms of changes in the way PyScaffold works should ideally be consolidated in the docs targeting developers/extension-writes.

So my conclusion here would be that we can use the wiki to document internal decisions (preferably linking to the relevant GitHub discussion). But I would prefer leaving "development guides" and any documentation about architecture in the website.

**Comment by julie777** [comment]
I agree with most of what you said above. I want the first docs a user sees to make them comfortable with the product. Make it easy for them to get started. 

I think that we could improve by
- making PyPI and github README send them to the docs as quickly as possible
- reorganizing the docs a bit to make getting started even easier. 
- I think that any architecture that an end user needs should be on the website

However, more and more I believe that all other doc should be on the wiki. It is easy to have links from the docs to the wiki, and as a user gets to the point where they want to understand how pysk works they will follow those links. 

The wiki makes it so easy to crosslink pages that small focused pages can become the norm, unlike the docs which have very long pages.

I am going to try and spend more time on my wiki example before we have our talk. Maybe you two could take a look at it Saturday, or Sunday before our talk.

Given that we all have history with using wikis, I wonder what you saw as the pace of change on the wiki. I know that in my experience the wiki worked because it was easy, simple to create content without spending time learning about formatting, you didn't need permission to add something or change something. People start making things better every day, with a spelling fix here and rewording an awkward sentence there, and adding a page based on something they discovered while researching for a project.

Again, I am not asking for an overnight switch, but just that we give it a try and see how it goes.

As for users that start to learn more about pysk and start using the wiki they should be allowed to contribute to the wiki. They will have tips for others, comments about how to do something, etc. (That reminds me, I have to check if Github supports page comments.)

**Reply by FlorianWilhelm** [reply]
Crosslinking is also easy in Sphinx. Also, our RTD Sphinx webpage looks really appealing compared to a Github wiki. Do you have an example for a project that extensively makes use of the Github wiki? 

Let's discuss this on Sunday but so far I guess we agreed to have additional dev docs in the wiki but not the whole user docs, right?

**Reply by FlorianWilhelm** [reply]
The wiki is now activated for your tests @julie777.

**Reply by julie777** [reply]
@FlorianWilhelm 

Thanks. I already have been working in my repo wiki. That should suffice until we make a decisions.
